Fix checking hash sum for JRE/JCE

vutny opened the pull request:
This PR brings back hash sum verifications for downloaded JRE tarball.
Unfortunately, Salt has had a bug, which caused complete ignorance of hash sum for local `file://` URI.
That should be fixed in upcoming Salt releases 2016.11.7 and 2017.7.0.
Related upstream PR is saltstack/salt#41914.
I propose to maintain a workaround for earlier versions.

@sroegner , please review.

vutny commented:
@noelmcloughlin Thanks for your review!

Items 1 and 2 are good improvements, despite they are little bit out of scope of this PR 😄 However, I'll do it.
Regarding item 3, I have constructed absolute path to JCE zip file to reuse it later. Please see my inline comment.

vutny commented on sun-java/jce.sls:
```diff
@@ -2,7 +2,7 @@
 {%- if java.jce_url is defined %}
 
-  {%- set zip_file = 'UnlimitedJCEPolicy.zip' -%}
+  {%- set zip_file = salt['file.join'](java.jre_lib_sec, 'UnlimitedJCEPolicy.zip') %}
```
@noelmcloughlin Here we have full path to the zip file, instead of relative one. This reduces code repetition and eliminates unnecessary `cwd` parameter in `download-jce-zip` state.

noelmcloughlin replied:
I should have noticed that - thanks.

vutny commented:
@noelmcloughlin I've made changes according to your kind suggestions. Please take a look.

noelmcloughlin commented:
@vutny, Doing a quick test on ubuntu. Please change string "sha256:" to "sha256=" in settings.sls. Thanks.

vutny commented:
@noelmcloughlin Have you grabbed all changed files? Which Salt version are you using on your Ubuntu box?
In `2016.11.6` the syntax for `file.check_hash` function requires exactly `sha256:****`. Please refer to the docs:
https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.file.html#salt.modules.file.check_hash

noelmcloughlin commented:
I am using salt 2016.11.6 (Carbon).  Using sha256:  threw an error but sha256= worked.  Let me test again in case I do not have all the changed files.

vutny commented:
Actually, it seems that both forms are valid, but I would like to follow official docs, because the `<hash_type>=<hash_value>` format could be eventually dropped off. It's confusing that `source_hash` parameter in other states expects this kind of syntax though... Bug, huh?

noelmcloughlin commented:
Its a doc bug?  Maybe <hash_type>:<hash_value> was never valid, and document should read <hash_type>=<hash_value> which is now depreciated in favour of  <hash_value>

vutny commented:
I guess you're right. This is the code from `file.check_hash` function:

```python
 731     for sep in (':', '='):
 732         if sep in file_hash:
 733             hash_type, hash_value = file_hash.split(sep, 1)
 734             break
```

And the function is used under the hood of both `file.managed` and `archive.extracted` states. To avoid even more confusion, I agree that it would be better to change `:` to `=`.

noelmcloughlin commented:
Yes, "=" worked for me - The documentation is wrong. Thanks @vutny

vutny commented:
@noelmcloughlin Fixed. Could you please merge?
